### PR TITLE
Re-written indent script.

### PR DESCRIPTION
**Pull request by lewis6991**
I've been working on the indent script over the last month and have completely re-written it to be more generalised and to support a lot of complex code indenting that was not possible with the last script. I've done quite a lot so I'll just summarise with points:
- Script no longer just looks at the last 2 lines to determine indenting. Now the script will search back through the file and stop when it detects a line it can deduce the indent from. 
- The main condition that the is looked for is for a bock start `{, (, begin`.
- The builtin function `searchpair` has been used to skip over nested blocks when searching back through the file.
- For the end of context blocks, e.g. `endfunction, endtask`, etc, the file is searched back for the first occurrence of the start of the context, e.g. `function void something();`. This has a great benefit of guaranteeing the correct indenting of these lines. In the old script, if a line was ever incorrectly indented, then the rest of the file would be incorrect as well. With the new script, incorrect indenting should be contained within context blocks.
- Comments are stripped out (in most cases) before they are analysed. This means inline comments can be placed anywhere and not mess up the indenting. e.g.

``` systemverilog
if (condition) /* comment */ begin /* comment */
    something();
    something2(); // The old script would of incorrectly de-indented this.
```
- Because the script is now analysing a lot more lines the performance is worse than the old script. However I did work on optimising this, and in its current state I think it is only about 2x slower which I think is acceptable.
- One-line blocks can now be nested! This was my main motivation for improving the script. This should also work for any kind of combination (see test). For example the following now works:

``` systemverilog
if (condition1)
    if (condition1)
        if (condition1)
            if (condition1)
                something();
something(); // De-indents!
```
- This new script should address all currently raised issues, plus lots of issues I never got round to raising. The test file is considerably larger.

I'll completely understand if you don't want to merge this in since it is a very significant change. However I will be continuing to work on this, so If you decide not to merge then I will continue work on my fork and will unfortunately not be able to carry on contributing to this repo as the codebase will diverge too quickly.

Despite this, let me know what you think and If you have any suggestions or changes you want to make to the PR, then they are welcome. :)

### CODE REVIEW

**Comment by vhda** [comment]
Hi Lewis,

I still did not fully review the code, but it really is simpler than the original code and way easier to read/update. Congratulations and thanks for this effort!

Nevertheless, I'd like to ask you to implement all the configuration variables that were available in the original version to guarantee that we are backwards compatible. I'm particularly concerned with `g:verilog_dont_deindent_eos` (although we should rename it to `b:` - in fact, I'll do that right away). If you feel that any of the options is no longer applicable, then please do update the indent script header and README.md accordingly.
Eventually I should update the indent test function to support both values of `b:verilog_dont_deindent_eos`... but that's not really important here.

I'm also worried about the performance of the new code. I've replaced the indent test case with a huge Verilog file and re-ran the test with the following results:
- original : ~7min
- new : ~14min

On some other smaller tests, there were situations where the new script was up to 3 times slower than the original one. So, even though I did not compare the results in terms of quality, the new indent script really takes a toll on performance.

Is there anything we can do to improve this? Maybe we could have more stop conditions, to avoid as much as possible reaching the top of the buffer while searching for context? (Again, I did not analyze the code in detail, so these are just random thoughts that may not even make sense :)

I still have to test it in a day-to-day usage scenario, where this might not be so noticeable. But I need the configuration variable mentioned above before I can do so.

**Comment by lewis6991** [comment]
Thanks for taking a look at the code.

Sorry about `b:verilog_dont_deindent_eos`, I forgot to implement that. Just pushed a commit which should work, though the change will not deindent any `);` whether it belongs to a module or not.

The performance is something that bothers me. The earlier version of the script was actually around 4-5x slower, however I was able to pick away at this by using `searchpair` and `search` as well as other things.

I think ultimately there has to be a trade off between quality and speed. Personally I would take quality as performance isn't really an issue as my main use for this script is for writing code. I rarely find myself in situations where I need to indent an entire file. The largest file size I deal with is upto 3000-4000 lines of code which I think this script will deal with fine. I don't know any use cases for this script on files larger than that.

Introducing more stop conditions will definitely work, however I can't think of any that will maintain the current quality, but this is something that can be worked on.

I did look into using:

```
map(synstack(line('.'), col('.'), 'synIDattr(v:val, "name")')
```

to determine context. This may increase performance, however I think the syntax file needs a lot of work for this to be properly explored.

Overall I think a 2-3x in performance loss isn't that bad when considering how more reliable the script is. I would never of run the old script on large files anyway because of how unreliable it was. If the old script ever got an indent wrong, then that propagated to the rest of the file. So really the comparison of performance of large files is pointless since the old script had a very low chance of indenting a large file correctly. Moreover, any indent script that is context aware will always be slower than one that is only evaluating 3 lines per indent (current, prev, prev2).

File: test/indent.sv [hunk 1]
```diff
@@ -166,13 +211,22 @@ class z;
...
+        end while (1);
+
     endfunction
 
     //function old_style_function_with_var(
     //    input a
     //);
     //reg test;
-    //begin 
+    //begin
```
Comment (vhda) [inline test/indent.sv:197]: This white space was here because I identified a bug where this white space would break indenting.
For this reason, I would like to keep it to make sure we do not reintroduce such a bug.

File: test/indent.sv [hunk 2]
```diff
@@ -252,22 +324,359 @@ device d1 (
     .port (port[1]),
     // .port1(), comment
     /**/.port2(), // comment
-    /*.port3(), */   
+    /*.port3(), */
```
Comment (vhda) [inline test/indent.sv]: Same as above.

File: test/indent.sv [hunk 3]
```diff
@@ -252,22 +324,359 @@ device d1 (
...
+        deindent_please();
+        dont_deindent_please();
+    endfunction : hello
+endclass : a
+
+module a;
+(
+    input clk,
+    output x
+);
```
Comment (vhda) [inline test/indent.sv:513]: Can a port list follow a module declaration ending in ';' ?

**Comment by NicoLugil** [comment]
I am following this thread with a lot of interest, and appreciate the time you guys are putting into this. 

I just would like to give my opinion on speed: my real interest is in doing a good job of indenting. Currently I need to write some stuff in particular ways that I dont particularly like to get a correct indent.
I have the habbit (from past C++ work) to almost continuously re-indent the whole file via a shortcut. So if something is broken it keeps getting re-introduced.

Most of my files are way smaller than 1000 lines of code, so I think speed is not that important for me. I rather have a good indentation that is somewhat slower. A factor of 2 to 3x slower isnt a big deal in my opinion. It would be the same a switching to a somewhat older server.

**Comment by vhda** [comment]
Hi Nico,

Could you merge this branch locally?
Feedback is very welcome at this time! So having you test it would be wonderful :)

**Comment by NicoLugil** [comment]
I would be happy to test this branch, but I am not en experienced git user, I don't know how I could get this version/branch so that I can download the files to my PC.

File: test/indent.sv (hunk 3 above)
Comment (lewis6991) [inline test/indent.sv:513]: Pretty sure it can. Its how you import packages.

```
module intf import some_pkg::*;
(
  // ports
);
endmodule
```

Looks weird I know, but that's SystemVerilog for you.

**Comment by lewis6991** [comment]
Probably easiest just to download my branch directly:

``` shell
cd <plugin directory>
git clone https://github.com/lewis6991/verilog_systemverilog.vim
```

If you use a package manager (such as plug or vundle) then place this (or similar) in your .vimrc:

```
call plug#begin('~/.vim/plugged')
Plug 'lewis6991/verilog_systemverilog.vim'
call plug#end()
```

File: test/indent.sv (hunk 1 above)
Comment (lewis6991) [inline test/indent.sv:197]: Yeah sorry, you've mentioned this before. I'll add some whitespace randomly to the test.

File: test/indent.sv (hunk 3 above)
Comment (vhda) [inline test/indent.sv:513]: Ok, so the `;` is from the `import` statement and is not part of the `module` declaration.
I think the test would be more correct to include that `import` statement in this line.
As far as I can see in the Formal Syntax definition of both Verilog and SystemVerilog it isn't possible for `module` to have a `;` after the module name.

**Comment by vhda** [comment]
This is how you add a new remote to your git repository:

```
git remote add lewis https://github.com/lewis6991/verilog_systemverilog.vim
git fetch lewis
git checkout -b test_new_indent lewis/master
```

File: test/indent.sv (hunk 3 above)
Comment (lewis6991) [inline test/indent.sv:513]: Your right. Just tested with my compiler. I've now changed the test accordingly.

**Comment by NicoLugil** [comment]
I am not a git user and we don't have git at the office, nor can I install it. I installed git at home, did the git clone https://github.com/lewis6991/verilog_systemverilog.vim and will move the downloaded files to the office to try.

Wish github would allow me to download the package from the site, but only the master branch seems to be selectable.

**Comment by lewis6991** [comment]
If you go to https://github.com/lewis6991/verilog_systemverilog.vim, you can download my master branch from there. Or download from https://github.com/lewis6991/verilog_systemverilog.vim/archive/master.zip

**Comment by vhda** [comment]
@lewis6991 I had some updates to the syntax for indenting purposes, which I have now uploaded into the following branch: https://github.com/vhda/verilog_systemverilog.vim/tree/indent_from_syntax

I had the intention to try to make some tests over your script, but I really do not seem to find the time to do that. So maybe you could try to use the new `verilogFunction` (and others) syntax areas to improve its performance?

**Comment by lewis6991** [comment]
Ok, I'll give it a go. However I'll still need to make more changes to the syntax file to create regions for `()`, `{}`, `case() endcase` and `fork join`.

**Comment by vhda** [comment]
Try focusing on the existing ones, for now, just to check if there are any performance improvements. If there aren't, then we probably don't gain anything with this solution.

**Comment by lewis6991** [comment]
I'm pretty sure it's required because the current script keeps searching upwards until it finds a stop condition. If we were to use the syntax checks, the stop condition would have to be when `l:oneline_mode = 0`.  There isn't really any other way that's obvious to me.

I'll think about it some more.

**Comment by NicoLugil** [comment]
I am trying out the version from https://github.com/lewis6991/verilog_systemverilog.vim/archive/master.zip

One thing I noticed is that the indenting of:

assert(W%2==0) else `uvm_fatal(get_typename(),$sformatf("W=%d is not an even number - was not what I expected",W))

isn't going as expected. In the old version neither by the way.

**Comment by lewis6991** [comment]
What do you expect it to do. I've had a look and it seems fine to me. Have you placed any newlines anywhere?

btw, if you place ``` around text on here, it will display it as code:

```systemverilog
assert(W%2==0) else `uvm_fatal(get_typename(),$sformatf("W=%d is not an even number - was not what I expected",W))
```

becomes:

``` systemverilog
assert(W%2==0) else `uvm_fatal(get_typename(),$sformatf("W=%d is not an even number - was not what I expected",W))
```

**Comment by NicoLugil** [comment]
the line following the assert gets an extra indent:

```
assert(x==0) else `uvm_fatal("x is not 0")
    y=6;
```

I would expect

```
assert(x==0) else `uvm_fatal("x is not 0")
y=6;
```

**Comment by lewis6991** [comment]
That works fine for me too. Are you sure you are using the updated script? Try and isolate the issue further by checking that other code above in the file isn't causing the issue.

**Comment by NicoLugil** [comment]
It are the only 2 lines in the file. Must be I am not using the updated script. My goal was to download https://github.com/lewis6991/verilog_systemverilog.vim/archive/master.zip and use those. I will give it another try.

**Comment by NicoLugil** [comment]
Ok, it is fine now, I unzipped the files in the wrong location orginally.

**Comment by lewis6991** [comment]
Let me know if you find any more issues :)

**Comment by NicoLugil** [comment]
I just wanted to comment that end else begin wasnt working fine, but I see you just fixed that :-)

One weird indent is the following code:

```
virtual class DataTypes_packer#(type T, int L, int W, int I);

        /*
        // packing functions
        extern static function void pack(const ref T in[], ref bit [L-1:0] out);
    extern static function void pack2(const ref t2#(W,I) in[], ref bit [L-1:0] out);
        */

        // unpack functions:
        extern static function void unpack(const ref bit [L-1:0] in, ref T out[]);
    extern static function void unpack5(const ref bit [L-1:0] in, ref t5#(W,I) out[], input int n_MSB2ignore=0);

endclass
```

**Comment by lewis6991** [comment]
That should fix it. :)

**Comment by NicoLugil** [comment]
Yes fixed :-)
I am very happy with the new script so far. I dont experience any extra latency or so.

Preprocessor statements aren't indenting as I would have expected:

```
`ifndef A
    `ifndef  B
        `define C
        ex ex_i(
    `else
            ex_G ex_G_i(
    `endif
`else
    ex_H ex_H_i(
`endif
.p_Clk(clk),
.p_ResetCfg(Reset)
    );
```

Maybe we have different expectations, but in my opinion, the `ifdef, etc... should not influence the indenting of non preprocessor statements

**Comment by lewis6991** [comment]
I think we do have different expectations. I can easily make this an option though.

**Comment by lewis6991** [comment]
Try that. Indentation of preprocessor statements is off by default. I may make the default on in the future though as this is what the old script did.

**Comment by NicoLugil** [comment]
this indents it now as follows:

```
`ifndef A
`ifndef  B
`define C
ex ex_i(
`else
    ex_G ex_G_i(
`endif
`else
        ex_H ex_H_i(
`endif
            .p_Clk(clk),
            .p_ResetCfg(Reset)
        );
```

I actually would have liked:

```
ex_H ex_H_i(
    .p_Clk(clk),
    .p_ResetCfg(Reset)
);


`ifndef A
`ifndef  B
`define C
ex ex_i(
`else
ex_G ex_G_i(
`endif
`else
ex_H ex_H_i(
`endif
   .p_Clk(clk),
   .p_ResetCfg(Reset)
);
```

so the same indenting as the top example for statement alternatives.

If too difficult, this is not so important.

BTW: how can I switch "Indentation of preprocessor statements", do I need to alter a file or is it via a Vim command?

**Comment by lewis6991** [comment]
At the moment you need to place this in your vimrc to enable indenting of preproc:

```
let b:verilog_indent_preproc=1
```

**Comment by lewis6991** [comment]
I think it would be impossible to implement what you are asking because the script doesn't know the values of the defines. The script will indent as if the preprocessor statements where invisible. So the following indentation would be correct:

``` systemverilog
ex ex_i(
    ex_G ex_G_i(
        ex_H ex_H_i(
            .p_Clk(clk),
            .p_ResetCfg(Reset)
        );
```

One potential work around would be to write the code as:

``` systemverilog
`ifndef A
`ifndef B
`define C
ex ex_i
`else
ex_G ex_G_i
`endif
`else
ex_H ex_H_i
`endif
(
    .p_Clk(clk),
    .p_ResetCfg(Reset)
);
```

**Comment by NicoLugil** [comment]
I think you are absolutely right, I was assuming that the script could in some way interpret the defines. Your suggestion with moving the bracket is fine!

**Comment by NicoLugil** [comment]
I think there might be an issue with the wait fork statement:

```
fork
    begin

        for(int i=0;i<5;i++) begin
            automatic int idx=i;
            fork
                begin
                    $display("%fns: %0d start",$realtime/1ns,idx);
                    my_seq[idx].go();
                end
            join_none
        end

        // wait for all forks to end
        wait fork;
            $display("%fns: all done",$realtime/1ns);
        end
        join

        $display("hi");
```

**Comment by vhda** [comment]
Tomorrow I will start a new task and will be able to (finally) start testing this Pull Request! :)
I have the bad habit of keeping the same Vim session while working on a specific task...

**Comment by lewis6991** [comment]
The last several commits where to correct lots of weird indent behaviour. Because of this performance has decreased another 5-20%.

However I have been working on another script on the side that uses the syntax file. I've made decent progress with that so far but it is still very much a work in progress.

**Comment by vhda** [comment]
I apologize for not have been of much help lately, but time has been short.

Are you able to see any significant performance improvements when using syntax? In my opinion, syntax (and in particular syntax folding) is one of the biggest shortcomings in Vim.

**Comment by lewis6991** [comment]
The performance improvement I'm seeing so far is like 3x than with this one. At the moment it gets most of the context from using the `synstack()` function. My difficulty now is trying to get indenting working properly for oneline statements. I looked at trying to leverage `cindent` but no luck so far.

**Comment by lewis6991** [comment]
Also the performance with the syntax script will be magnitudes quicker if the file has a very large module or class.

**Comment by vhda** [comment]
I think that for one liners the best approach would be to use a `getc()` kind of approach instead of regular expressions. I use such an approach in the omni-completion script and I have to admit: the code is by no means easy to read.

**Comment by lewis6991** [comment]
Where is the documentation for `getc()`?

**Comment by vhda** [comment]
I was referring to the `getc()` from C language :P 
Should have made that clear, sry

**Comment by lewis6991** [comment]
If that route was taken, wouldn't that then require the user to compile to plugin?

**Comment by lewis6991** [comment]
Oh, sorry. You meant a `getc` kind of approach. I get what you are saying.

**Comment by vhda** [comment]
lol
I was already in the middle of a long description of what I meant. Good thing github dynamically updates the thread! :D

**Comment by lewis6991** [comment]
The main problem with oneline statements is that the statement can be a block. e.g.

``` systemverilog
if (1)
    if (1)
        if (1)
        begin
            statement(); // synstack can't be used much here.
        end
```

In these cases we can determine that the statement is in a block using `synstack` but then we would need to search up the file until we reach `begin` to get the right indent. This isn't too bad really, I just need to find a chunk of time to get it done.

**Comment by vhda** [comment]
Ok, I did not understand what you meant by "oneline statements" then. But I can now see that the syntax approach does not work in all scenarios.

It would be great if we could somehow have invisible "marks" that we could use for this purpose. This way we would only need to navigate until the closest "mark" of the type we want to search for and make a decision from there.

**Comment by vhda** [comment]
The `verilog_dont_deindent_eos` option appears to not be working. Could you please double check on your side?

**Comment by lewis6991** [comment]
Fixed. There was a missing `\` on an `if`.

**Comment by lewis6991** [comment]
@vhda checkout my other branch `syntax_indent_dev` and tell me what you think of the indent script in terms of performance. The code isn't pretty, but it mostly provides the same indenting as this script and should be a lot quicker on files with large modules and classes. With that said it is actually slower on `test/indent.sv` but that's only because that file has very small classes and modules and it indents quickly anyway.

In order for this script to work I've had to add a lot more `syn regions` to the syntax file. This may have its own performance hit.

**Comment by vhda** [comment]
The `verilog_dont_deindent_eos` is fixed. Thanks.

Using the syntax based indenting on an unusually big Verilog file does not work. I gave up waiting for it after 2h30! It is possible that there is a loop somewhere in there, so I'll try to run it in debug mode to see if I find anything.

But the most probable scenario is that the syntax based approach is slower than simply analyzing the buffer contents.

**Comment by vhda** [comment]
I've been using the command `time make` to measure the performance of the indenting script, but this weekend I noticed that if the indent function is modified to run `gg=G` it takes much less time than calling `==` once per line. In fact, when using the later the time the new script takes to indent a huge file (while not analyzing the correctness) is the same as the original one. So, I'm not sure that it is correct to use the indentation test for performance measurement. I've submitted a "performance" target to the Makefile such that you can review this yourself.

Using the "performance" target I only see a ~4% performance improvement when using the syntax approach, so let's look into this in more detail at a later time.

In a daily use I am unable to detect any performance penalty when compared to the original script and, as such, I want to merge this Pull Request. But before doing so I need you to do the following changes:
- Rebase your changes over my master branch.
- Add profile target help description in the Makefile.
- Don't modify code in the indent test between 'Code from:' and 'End of copied code'.

There an issue when `g:verilog_dont_indent_eos` is set and we have an empty port list:

``` verilog
function name(
);
```

**Comment by lewis6991** [comment]
Thanks for reviewing. I'll try and make those changes as soon as possible.

It seems Vim is more efficient at indenting multiple lines, as opposed to one at a time. I think `g==G` is more reflective of practical use and as such would make a better method for testing.

Sorry for editing the indent test in the copied sections. Is it ok for me to leave it as it is or would you like me to revert the additions I made.

As for the `g:verilog_dont_indent_eos` issue, can you specify what indent you are seeing and what indent you expect? It seems to work fine for me, but as I don't personally use this setting we may have different expectations.

**Comment by vhda** [comment]
Take your time, as I also don't have that much free time nowadays :)

I did some testing but gave up before reaching any conclusions. What it looked like was that somehow Vim was indenting multiple lines for each `==` call. I'm honestly curious to understand what is it doing to take so much time. Will probably look into this again in the future, if I don't forget in the meanwhile.

My suggestion regarding the indent test case is that you use `git rebase -i` to re-edit those changes in order to move the code from the "pragma" zone into another area of the file.

What happens is that when I start creating a function I do the following:

``` verilog
function name(
);
```

and only after that will I add the port list. But with `g:verilog_dont_indent_eos` enabled I would expect the following:

``` verilog
function name(
    );
```

Please note that if I add the pin list and afterwards indent with `==` it has the expected behavior. And I just remembered that I also have `b:verilog_indent_modules` set, which might explain this behavior.

**Comment by lewis6991** [comment]
That should do it.

**Comment by vhda** [comment]
I would really appreciate if you could use `git rebase -i` to review the changes to these tests in the original commits. This is important to guarantee that `git blame` correctly reports the original commits in these lines. I understand that you may feel that this is not that important, but I really think we should always follow these flows.

I don't know how comfortable you are with this command, but using this as an example you can start the interactive rebase with `git rebase -i 732dcab48cd1c29b3cd18375d790ba60f033ba69^`. Then modify the contents of the rebase file as following (note the "edit" entries):

```
edit 732dcab Fix for issue 41.
pick df0487d Indent: Improved support for always blocks.
pick f1f3ded Indent: Fix for join keyword
pick 632d3d0 Indent: More simplifcations and refactoring.
pick 77f0efc Indent: Added file with code examples that fail indentation.
pick 99cb727 Indent: Added more unsupported examples.
pick 9aa5b8b Indent: Added support for local static functions.
pick f2597d5 Indent: Renamed function.
edit 1f910e2 Indent: Completely rewritten the script.
pick 00937bf Indent: Differentiate between different block types.
pick 1c582ed Indent: 25%+ speed increase.
pick c5c0516 Indent: Added expandtab to modeline to fix travis.
pick 95e329a Indent: Added scope to methods.
pick e189d20 Test: Moved functions into separate file.
pick 86d49d5 Indent: Added example of issue #43 to test.
pick a30de76 Indent: Fix for covergroups and added example to test.
pick bfab2eb Indent: Fix for sv packages and added example to test.
pick 131e8ba Indent: Added support for 'typedef class'.
pick b159d8a Indent: Removed else from indent keys.
pick e58286c Indent: Fixed typo.
pick ba77390 Indent: Support for 'b:verilog_dont_deindent_eos'
pick 76212d6 Test: Added trailing whitespace to indent test.
pick 4e2dde5 Test: Change module declaration to include package import.
pick 0074bc2 Indent: Fixed bug where indent was not correct whilst typing some specific indent keys.
pick 70e4856 Indent: Several corner case fixes.
pick e25e8c1 Indent: Better support for 'end else begin'.
pick 4547a92 Indent: Handle block comments better.
pick 8dad3d7 Indent: Added option for indenting preprocessor statements.
pick dda0a63 Indent: Support for nested modules.
pick aacb313 Indent: Fixed issue with wait_fork.
pick 4140674 Indent: Fix for closing parenthesis.
pick 32b0f80 Indent: Fix indent after an `endif.
pick 40e22d7 Indent: Fix for old style functions.
pick ad6cd4b Indent: Fix for b:verilog_dont_deindent_eos
pick a475347 Indent: Refactoring.
pick 6dabd8e Indent: Fix for 'end else'.
pick 388acc7 Test: Removed changes to copied test code.
pick d7deef4 Indent: Fix issue 51. Make script case sensitive.
```

After saving and exiting you can copy back the changes you want to remove using the command `git checkout -p 388acc7 -- test/indent.sv`. This will ask you interactively which changes you want copied back (in this case you only want to say yes to the 3rd patch).

You can then continue the rebase process with `git rebase --continue`, which will open a new editor and allow the commit update (you can use this opportunity to also add the missing commit type to the summary description).

Then you go to the second "edit" entry, where you run `git checkout -p 388acc7 -- test/indent.sv` again, but this time you want to say "yes" to the 2nd and 3rd patches. Again, continue the rebase with `git rebase --continue`.

At this point you will need to resolve a conflict. While you could resolve the conflict, let's keep it simple and import the necessary changes again. Start with `git reset test/indent.sv` and `git add -u`, followed by the usual `git checkout -p 388acc7 -- test/indent.sv`. Accept the 2nd patch and edit the 11th (by pressing "e"). Here you want to keep only the following "+" lines:

```
+task run_phase(uvm_phase phase);
+    int var0 = var1 +
+        var2 *
+        var3;
+
+    int var0 =
+        var1;
+
+    if (map.first(s))
+        do
+            $display("%s : %d\n", s, map[s]);
+        while (map.next(s));
+
+    label : assert(my_seq.randomize());
+    my_seq.start(low_sequencer_h);
+
+    assert(my_seq.randomize() with {Nr==6;});
+    my_seq.start(low_sequencer_h);
+
+    label : assert(my_seq.randomize() with
+        {Nr==6; Time==8;});
+    my_seq.start(low_sequencer_h);
+
+    assert(
+        my_seq.randomize() with
+    );
+
+    //task
+endtask
+
+function void sink_driver::build_phase(uvm_phase phase);
+
+    assert property (prop1)
+    else `uvm_fatal("TAG", "Assertion failed.")
+
+    do_something();
+
+    assert property (prop1)
+    else
+        `uvm_fatal("TAG", "Assertion failed.")
+
+    do_something();
+
+    if (condition)
+        something();
+    else `uvm_fatal("TAG", "This is invalid.")
+
+    do_something();
+
+    if (condition)
+        something();
+    else
+        `uvm_fatal("TAG", "This is invalid.")
+
+    do_something();
+endfunction
+
```

More conflicts follow, but I suggest that you use `git mergetool` in those as they are quite simple to resolve. I've followed these steps to avoid those changes (and others I noticed in the meanwhile) to the test case and you can see the results in my "lewis-master" branch. 

---

Ideally, you would simply squash these changes into a smaller number of independent commits in a separate branch and create a new PR. This is typically the preferred methodology and is widely used in open source projects (e.g. Linux Kernel and Git). Smaller and more consistent commits allow an easier code review. For example, commit 980c5417633380a1fcd308f8acc35fda88539ff3 could be squashed directly into 78e2c8462ae65e34b02412a666df68383ef4780b!

Just while reviewing the test case alone I saw lines being modified and then reverted in a later commit. This makes it very hard to keep track of what really changed and why!

For now I would like to keep commit d49e94e1262d02f490ed4b47ff6c6c48489fccb6 out of this PR. I feel that there may be another approach to that problem, but would like to merge everything else as soon as possible. (Want to start hacking the new script ;)

One last question, would you agree that keeping the original indentation function's would make it more coeherent in respect to the remainder of the plugin?

**Comment by lewis6991** [comment]
I've cleaned up the commits and squashed a few too. I've also dropped the last commit and saved it on a different branch. Are there any other changes or other commits I should squash?

I don't quite understand your question about the original indentation functions in regards to coherency. Could you elaborate?

**Comment by vhda** [comment]
The plugin is called verilog_systemverilog (at least for now;) so I think that it would make more sense to keep the indentation function named as `GetVerilog_SystemVerilogIndent()`.

But I already have some exceptions to this rule: configuration variables are named as `.:verilog_.*` to avoid huge settings in the user's vimrc. I'd like to avoid another incoherency by adding a function with the prefix `SystemVerilog`.

What do you think?

**Comment by lewis6991** [comment]
Thats fair enough. I'll make the change.

I must say though I do find the name `verilog_systemverilog` kind of redundant as one implies the other. In my eyes they're not really separate languages.

**Comment by vhda** [comment]
Finally we merge this excellent work. Thanks a lot Lewis!
